[cmd/agent] During Logs Agent startup, wait for the AutoConfig to run before starting ContainerCollectAll.

remeh opened the pull request:
### What does this PR do?

The ContainerCollectAll virtual configuration of the Logs Agent can start before any configuration file is processed, meaning that it would reap the tailing ownership of every running containers, even if a configuration file has been created through a ConfigMap to tail some containers.

This PR delays the introduction of the ContainerCollectAll source until the AutoConfig has been executed.

### Additional Notes

Some refactoring has been done because the main logs package contained some global variables which had to be contained in a more precise package.

### Describe your test plan

In a k8s cluster, enable the `logs_config.container_collect_all` flag and use a ConfigMap to add a configuration file tailing logs of a container. Validate that the logs from the configuration file is properly tailed by the custom config.
Feel free to ping me for questions.

remeh committed:
autoconfig: store in an atomic int when the AutoConfig has been executed.

remeh committed:
cmd/agent: while starting the Logs Agent, wait for the AutoConfig to run before starting ContainerCollectAll.

The ContainerCollectAll virtual configuration of the Logs Agent can start before
any configuration file is processed, meaning that it would reap the tailing ownership
of every running containers, even if a configuration file has been created to tail
some containers.

This PR delays the introduction of the ContainerCollectAll source until the AutoConfig
has been executed.

Some refactoring has been done because the main logs package contained some global
variables which had to be contained in a more precise package.

remeh committed:
cmd/common/agent: fix and add unit tests

ahmed-mez commented on pkg/logs/logs.go:
```diff
@@ -96,10 +83,15 @@ func Start() error {
+	// add the default sources after the AutoConfig has been runned once because
+	// we don't want the container_collect_all to reap all containers ownership
+	go func() {
+		common.BlockUntilAutoConfigRunnedOnce(time.Second * 30)
```
I'm afraid 30 seconds could be a lot, we should think about short lived containers that we could miss in 30 seconds. Could we make t configurable at least?

remeh replied:
Interesting.
This is a timeout because I don't want to have a `for {}` running forever in the case of an issue (bug) in the autoconfig.
In the normal behavior of an Agent startup, this timeout won't occur and this will probably be blocking only for hundreds of milliseconds (the time needed for the autoconfig to run). With this in mind, do you think this timeout should still be configurable? I've nothing against a new configuration field (except from the fact that the Agent has already a lot ;D), I just want to be sure that these 30s are properly understood.

ahmed-mez replied:
yes we're on the same page, of course we should keep the timeout as protection. I'm just saying 30s as default timeout could be a lot in case of an issue, if we know that the nominal case should only takes < 1 second I guess we can set a 5s timeout no?
Taking into consideration this is only executed at the agent start/restart, it's fait to say it's not very likely to have the timeout reached + a short lived container existed at the same time but theoretically it could happen. Having a shorter timeout would reduce this risk considerably.

L3n41c replied:
I’m also wondering under which conditions the timeout will be hit.
Right now, if the AutoConfig hasn’t run once before the timeout, we continue as if it successfully ran (And we log a warning)
I’m wondering what’s best between continuing in best effort mode (ignoring the fact the AutoConfig didn’t ran) and making the issue “more visible”.

The reason why I’m not a big fan of too aggressive timeouts is because in cases where the agent run slower than usual (because it is running on a heavily loaded host or because it is running with instrumentation features that are slowing it down (like valgrind)), the agent will not run slower, it will also behave differently.

remeh replied:
First, @L3n41c you made me realize that throwing an error is better than throwing a warning here, I'll change this, it'll be a bit more visible. Maybe we should (in another PR) add the autoconfig in the health mechanism of the Agent, WDYT?

As mentioned by @L3n41c this lenient timeout is to give some space to the AutoConfig to run on loaded hosts. I guess using a configurable field won't hurt anyways, let me add that.

remeh replied:
Also I should mention: even if the AutoConfig takes longer, it won't stop it for loading, it'll just start doing this (loading some default sources in the logs agent) before the AutoConfig has finished to prepare.

ahmed-mez commented on pkg/autodiscovery/autoconfig.go:
```diff
@@ -65,6 +66,8 @@ type AutoConfig struct {
 	store              *store
 	m                  sync.RWMutex
+	// runnedOnce is an atomic uint32 set to 1 once the AutoConfig has been executed
+	runnedOnce uint32
```
nit
```suggestion
	ranOnce uint32
```

remeh replied:
🙇

ahmed-mez commented on cmd/agent/common/autoconfig.go:
```diff
@@ -107,3 +109,19 @@ func SetupAutoConfig(confdPath string) {
+
+// BlockUntilAutoConfigRunnedOnce blocks until the AutoConfig has been runned once.
+// It also returns after the given timeout.
+func BlockUntilAutoConfigRunnedOnce(timeout time.Duration) {
```
nit
```suggestion
func BlockUntilAutoConfigRanOnce(timeout time.Duration) {
```

remeh replied:
🙇

ahmed-mez commented on pkg/logs/logs.go:
```diff
@@ -96,10 +83,15 @@ func Start() error {
-	for _, source := range config.DefaultSources() {
-		sources.AddSource(source)
-	}
+	// add the default sources after the AutoConfig has been runned once because
+	// we don't want the container_collect_all to reap all containers ownership
```
wondering if we could test this behaviour with an integration test

ahmed-mez reviewed (commented):
Thanks for the fix, left few questions.
I believe both core and containers should QA this change so I'm adding the containers team label.

remeh committed:
Spelling.

L3n41c commented on cmd/agent/common/autoconfig.go:
```diff
@@ -107,3 +109,19 @@ func SetupAutoConfig(confdPath string) {
+
+// BlockUntilAutoConfigRanOnce blocks until the AutoConfig has been ran once.
+// It also returns after the given timeout.
+func BlockUntilAutoConfigRanOnce(timeout time.Duration) {
+	now := time.Now()
+	for {
+		time.Sleep(100 * time.Millisecond) // don't hog the CPU
+		if AC.HasRanOnce() {
+			return
+		}
+		if time.Since(now) > timeout {
+			log.Warn("BlockUntilAutoConfigRanOnce timeout after", timeout)
+			return
+		}
+	}
+}
```
It looks like the goal of this function is to wait for a condition to be met (`AC.ranOnce` to be set to 1).
May be [`sync.Cond`](https://golang.org/pkg/sync/#Cond) could advantageously replace the use of `time.Sleep`.

[`sync.Cond`](https://golang.org/pkg/sync/#Cond) is specifically crafted to make a goroutine sleep until a given condition is met.

remeh replied:
Good point! I'll do the change.

remeh replied:
Actually using `sync.Cond` would be difficult for some reasons:
  - the part waiting for the block will still have to loop forever until the AC is instanciated and its `Cond` is instanciated and ready as well.
  - sharing the `sync.Cond` in this case is kinda painful because of some global variables of the AC.
Do you mind if I keep this implementation or would you really prefer for me to push forward with `sync.Cond`? I think the trade-off of this sleep brings some simplicity in the code + the capacity to timeout, which I would not able to do with a `sync.Cond`.

remeh committed:
cmd/agent: autoconfig should send an error instead of a warning when it is long to startup.

remeh committed:
logs: use a configurable field for the maximum delay before introducing default sources.

ahmed-mez commented on pkg/logs/logs.go:
```diff
@@ -86,7 +87,7 @@ func Start() error {
 	// we don't want the container_collect_all to reap all containers ownership
 	go func() {
-		common.BlockUntilAutoConfigRanOnce(time.Second * 30)
+		common.BlockUntilAutoConfigRanOnce(time.Millisecond * time.Duration(coreConfig.Datadog.GetInt("ac_load_timeout")))
```
we can use `Datadog.GetDuration` here
```suggestion
		common.BlockUntilAutoConfigRanOnce(time.Millisecond * coreConfig.Datadog.GetDuration("ac_load_timeout"))
```

remeh replied:
🙇‍♂️

remeh committed:
logs: use config.GetDuration instead of config.GetInt to get the duration.

remeh committed:
Update pkg/logs/logs.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

olivielpeau commented on pkg/logs/logs.go:
```diff
@@ -96,10 +84,15 @@ func Start() error {
+	go func() {
+		common.BlockUntilAutoConfigRanOnce(time.Millisecond * coreConfig.Datadog.GetDuration("ac_load_timeout"))
+		log.Debug("Adding DefaultSources to the Logs Agent")
+		for _, source := range config.DefaultSources() {
```
Does it make sense to wait for _all_ default sources? (I think there can be 2 right now: "ContainerCollectAll" and the snmp trap source).

Typically I don't think we need to wait for the snmp trap source (I think waiting could actually make us miss some snmp traps while Autodiscovery finishes its first collection). To avoid this I'd be OK with separating the setup of the ContainerCollectAll source (async) from the snmp traps source (sequential in `Start`), wdyt?

remeh replied:
My last commits should have addressed all your feedback.

olivielpeau commented on pkg/logs/logs.go:
```diff
@@ -96,10 +84,15 @@ func Start() error {
+	// add the default sources after the AutoConfig has been executed once because
+	// we don't want the container_collect_all to reap all containers ownership
+	go func() {
+		common.BlockUntilAutoConfigRanOnce(time.Millisecond * coreConfig.Datadog.GetDuration("ac_load_timeout"))
```
I think `GetDuration` expects the config value to be set to a string formatted for [`time.ParseDuration`](https://golang.org/pkg/time/#ParseDuration).

In general, in config, I prefer integers in a defined unit over a string duration, so that users can only set values that make sense (not `1ns`), but since this is really an internal config option I don't mind. If we go with durations let's update the default value in `config.go` to a string

olivielpeau replied:
nit: no need to block and wait if there are no default sources at all

olivielpeau reviewed (commented):
Thanks! Made a couple of comments.

To summarize, it looks like the current design of ContainerCollectAll as a whole forces us to make a tradeoff between:

1. missing short-lived containers that we'd want to tail with ContainerCollectAll, if they come up and down during autoconfig's first run.
2. entirely missing log integration configurations scoped to a container when that container is picked up by ContainerCollectAll before its log integration configs are passed

I think this PR looks like the better tradeoff to me. But long-term we'll want to rework ContainerCollectAll.

remeh committed:
Revert "logs: use config.GetDuration instead of config.GetInt to get the duration."

This reverts commit 06d37ebf97260edf9d92f47c0d9f20110b4e141d.

remeh committed:
logs: start the SNMPTrapsSource immediately if enabled (don't wait for the AC to init).

remeh committed:
Merge branch 'remeh/logs-delay-collect-all' of github.com:DataDog/datadog-agent into remeh/logs-delay-collect-all

remeh committed:
logs: fix unit tests following `DefaultSources()` changes.

gaetan-deputier commented on pkg/logs/status/status.go:
```diff
@@ -13,10 +13,23 @@ import (
 	builder  *Builder
 	warnings *config.Messages
 	errors   *config.Messages
+
```
Thanks for adding that!

olivielpeau commented on pkg/logs/logs.go:
```diff
@@ -9,7 +9,10 @@ import (
 	"sync/atomic"
+	"time"
 
+	"github.com/DataDog/datadog-agent/cmd/agent/common"
```
I didn't notice on first review, but having to import a package from `cmd/` in `pkg/logs/` is a big code smell.

It looks like it's not the only place in `pkg/` where we import `cmd/`, but still better to avoid this: we're introducing a dependency direction we shouldn't have IMO.

Instead, the `logs` pkg could expose a new exported function that autoconfig would call once it's finished its first run, WDYT?

remeh replied:
Good catch. I've pushed a commit which should avoid this by using an anonymous function to retrieve the AC instance.

remeh committed:
logs: move BlockUntilAutoConfigRanOnce from `cmd/agent/common` -> `pkg/logs`

olivielpeau commented on pkg/logs/logs.go:
```diff
@@ -96,14 +87,44 @@ func Start() error {
+		if getAC() == nil {
+			continue
+		}
+		if getAC().HasRanOnce() {
```
```suggestion
		if getAC().HasRunOnce() {
```

olivielpeau commented on pkg/logs/logs.go:
```diff
@@ -96,14 +87,44 @@ func Start() error {
 	return nil
 }
 
+// BlockUntilAutoConfigRanOnce blocks until the AutoConfig has been ran once.
```
```suggestion
// BlockUntilAutoConfigRanOnce blocks until the AutoConfig has been run once.
```

olivielpeau commented on pkg/autodiscovery/autoconfig.go:
```diff
@@ -190,6 +193,16 @@ func (ac *AutoConfig) AddConfigProvider(provider providers.ConfigProvider, shoul
+	log.Debug("LoadAndRun done.")
+}
+
+// HasRanOnce returns true if the AutoConfig has ran once.
```
```suggestion
// HasRunOnce returns true if the AutoConfig has ran once.
```

olivielpeau commented on pkg/autodiscovery/autoconfig.go:
```diff
@@ -190,6 +193,16 @@ func (ac *AutoConfig) AddConfigProvider(provider providers.ConfigProvider, shoul
+
+// HasRanOnce returns true if the AutoConfig has ran once.
+func (ac *AutoConfig) HasRanOnce() bool {
+	if ac == nil {
+		return false
+	}
```
I think this check is useless? (if it's nil we'd get a nil pointer dereference when calling this)

remeh replied:
It is useful in the case of somebody calling `HasRunOnce` too soon on a nil instance of the AutConfig, because no we can call a method on a nil instance of something, this would not crash. Let's discuss this offline.

olivielpeau replied:
you're right, I stand corrected, thanks for clarifying 👍

olivielpeau commented on pkg/autodiscovery/autoconfig.go:
```diff
@@ -190,6 +193,16 @@ func (ac *AutoConfig) AddConfigProvider(provider providers.ConfigProvider, shoul
+}
+
+// HasRanOnce returns true if the AutoConfig has ran once.
+func (ac *AutoConfig) HasRanOnce() bool {
```
```suggestion
func (ac *AutoConfig) HasRunOnce() bool {
```

olivielpeau reviewed (approved):
Thanks, only left a few nits inline, other than that LGTM!

remeh committed:
Apply suggestions from code review

Co-authored-by: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

remeh committed:
logs: remove unused nil AC testing while starting the Logs Agent.

It was actually avoiding the timeout to trigger if the AutoConfig
is never instanciated.
